scheduler/cloudformation/template.go

```diff
@@ -0,0 +1,281 @@
+	// should bind to. Tihs value is also exposed to the container through the PORT
+	// environment variable.
+	ContainerPort = 8080
+
+	schemeInternal = "internal"
+	schemeExternal = "internet-facing"
+
+	defaultConnectionDrainingTimeout int64 = 30
+	defaultCNAMETTL                        = 60
+)
+
+// This implements the Template interface to create a suitable CloudFormation
+// template for an Empire app.
+type EmpireTemplate struct {
+	// The ECS cluster to run the services in.
+	Cluster string
+
+	// The hosted zone to add CNAME's to.
+	HostedZone *route53.HostedZone
+
+	// The ID of the security group to assign to internal load balancers.
+	InternalSecurityGroupID string
+
+	// The ID of the security group to assign to external load balancers.
+	ExternalSecurityGroupID string
+
+	// The Subnet IDs to assign when creating internal load balancers.
+	InternalSubnetIDs []string
+
+	// The Subnet IDs to assign when creating external load balancers.
+	ExternalSubnetIDs []string
+
+	LogConfiguration *ecs.LogConfiguration
+}
+
+// Execute builds the template, and writes it to w.
+func (t *EmpireTemplate) Execute(w io.Writer, data interface{}) error {
+	v, err := t.Build(data.(*scheduler.App))
+	if err != nil {
+		return err
+	}
+
+	raw, err := json.MarshalIndent(v, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	_, err = io.Copy(w, bytes.NewReader(raw))
+	return err
+}
+
+// Build builds a Go representation of a CloudFormation template for the app.
+func (t *EmpireTemplate) Build(app *scheduler.App) (interface{}, error) {
+	parameters := map[string]interface{}{}
+	resources := map[string]interface{}{}
+	outputs := map[string]interface{}{}
+
+	serviceRole := "ServiceRole" // TODO: Build a service role.
+	parameters["ServiceRole"] = map[string]string{
+		"Type":    "String",
+		"Default": "ecsServiceRole",
+	}
+
+	for _, p := range app.Processes {
+		// CloudFormation only allows alphanumeric resource names, so we
+		// have to normalize it.
+		r := regexp.MustCompile("[^a-zA-Z0-9]")
+		key := r.ReplaceAllString(p.Type, "")
+
+		ulimits := []map[string]interface{}{}
+		if p.Nproc != 0 {
+			ulimits = []map[string]interface{}{
+				map[string]interface{}{
+					"Name":      "nproc",
+					"SoftLimit": p.Nproc,
+					"HardLimit": p.Nproc,
+				},
+			}
+		}
+
+		portMappings := []map[string]int64{}
+
+		labels := p.Labels
+		if labels == nil {
+			labels = make(map[string]string)
+		}
+
+		environment := []map[string]string{}
+		for k, v := range p.Env {
+			environment = append(environment, map[string]string{
+				"Name":  k,
+				"Value": v,
+			})
+		}
+
+		loadBalancers := []map[string]interface{}{}
+		if p.Exposure != nil {
+			scheme := schemeInternal
+			sg := t.InternalSecurityGroupID
+			subnets := t.InternalSubnetIDs
+
+			if p.Exposure.External {
+				scheme = schemeExternal
+				sg = t.ExternalSecurityGroupID
+				subnets = t.ExternalSubnetIDs
+			}
+
+			instancePort := int64(9000) // TODO: Allocate a port
```

Comment:
One thing that is a little more challenging in a CloudFormation world is instance port allocation. Because we no longer have direct control over creating and deleting load balancers, it's hard to allocate and release ports for them.

I was thinking that we could just add a custom resource with a lambda function that allocates allocates and releases ports, so it's just native CloudFormation (and can be re-used outside of Empire).